src/cargo/sources/path.rs

```diff
@@ -411,6 +470,228 @@ impl<'gctx> PathSource<'gctx> {
+    /// traversing the git working tree, while avoiding ignored files.
+    ///
+    /// This looks into Git submodules as well, resolving them to individual files.
+    /// Symlinks to directories will also be resolved, which may include `.git` folders
+    /// in entirety if the link points to a submodule.
+    fn list_files_gix(
+        &self,
+        pkg: &Package,
+        repo: &gix::Repository,
+        filter: &dyn Fn(&Path, bool) -> bool,
+    ) -> CargoResult<Vec<PathBuf>> {
+        use gix::dir::entry::Status;
+        warn!("list_files_gix {}", pkg.package_id());
+        let options = repo
+            .dirwalk_options()?
+            .emit_untracked(gix::dir::walk::EmissionMode::Matching)
+            .emit_ignored(None)
+            .emit_tracked(true)
+            .recurse_repositories(false)
+            .symlinks_to_directories_are_ignored_like_directories(true)
+            .emit_empty_directories(false);
+        let index = repo.index_or_empty()?;
+        let root = repo
+            .work_dir()
+            .ok_or_else(|| anyhow::format_err!("can't list files on a bare repository"))?;
+        assert!(
+            root.is_absolute(),
+            "BUG: paths used internally are absolute, and the repo inherits that"
+        );
+
+        let pkg_path = pkg.root();
+        let mut target_prefix;
+        let repo_relative_pkg_path = pkg_path.strip_prefix(root).ok().unwrap_or(Path::new(""));
+
+        let pathspec = {
+            let mut include = gix::path::to_unix_separators_on_windows(gix::path::into_bstr(
+                repo_relative_pkg_path,
+            ));
+
+            target_prefix = include.clone();
+            target_prefix.to_mut().push_str(if include.is_empty() {
+                "target/"
+            } else {
+                "/target/"
+            });
+
+            // The `target` directory is never included if it is in the package root.
+            let exclude = {
+                let mut pattern = include.clone();
+                pattern
+                    .to_mut()
+                    .insert_str(0, if include.is_empty() { ":!" } else { ":!/" });
+                pattern.to_mut().push_str(b"/target/");
+                pattern
+            };
+            include.to_mut().insert_str(0, ":/");
+
+            vec![include, exclude]
+        };
+
+        let mut delegate = Delegate::new(self, pkg, pkg_path, repo, root, filter)?;
+        repo.dirwalk(
+            &index,
+            pathspec,
+            &Default::default(),
+            options,
+            &mut delegate,
+        )?;
+
+        let mut files = delegate.into_result()?;
+        // Append all normal files that might be tracked in `target/`.
+        files.extend(
+            index
+                .prefixed_entries(target_prefix.as_ref())
+                .unwrap_or_default()
+                .iter()
+                .filter(|entry| {
+                    entry.stage() == 0
```

Comment:
What does `stage() == 0` mean?

Reply:
In a later commit this receives a comment that makes it a bit clearer.

```
// probably not needed as conflicts prevent this to run, but let's be explicit.
```

Stage zero essentially excludes entries that are related to conflicts, and it's something the code calling this would prevent as it checks for changes first. However, if that would be bypassed for any reason, it's better not to risk adding entries that point to the same path, but at different stages.

Reply:
Why are numeric codes being used for this?

Reply:
I see what you mean now, and investigated this. 
First of all, the stage as it is here is already quite accessible compared to [decoding it by hand from `flags`](https://docs.rs/git2/latest/git2/struct.IndexEntry.html#structfield.flags) as is done in `libgit2`. But then I was wondering why it's not an enum and [found the `stage()` code](https://github.com/Byron/gitoxide/blob/27471e7f421446cc8f6d4543bdf85fe4e753b944/gix-index/src/entry/flags.rs#L64-L66) which initially looked like this bitshift could yield many different values. Digging deeper, it shows this was wrong and the mask is truly 2 bits in size, so an `enum` with 4 variants would be much better.

I pushed the (breaking) changes [into this pending PR](https://github.com/Byron/gitoxide/pull/1317) so that when that version releases and Cargo is updated, it will see the new enums and use them.

Thus, this can't change immediately (to not cause avoidable version-churn), but it will trickle through for sure.